File: leetcode_804/solution.js
/**
 * @param {string[]} words
 * @return {number}
 */
var uniqueMorseRepresentations = function(words) {
    const codes = [
        '.-', '-...', '-.-.', '-..', '.', '..-.', '--.', '....', '..', '.---', '-.-', '.-..', '--', 
        '-.', '---', '.--.', '--.-', '.-.', '...', '-', '..-', '...-', '.--', '-..-', '-.--', '--..'
    ];
    const morseIndex = char => char.charCodeAt(0) - 'a'.charCodeAt(0);
    return new Set(words.map(
        word => [...word].map(char => codes[morseIndex(char)]).join('')
    )).size;
};

function main () {
    words = ["gin", "zen", "gig", "msg"];
    ret = uniqueMorseRepresentations(words);
    console.log(ret);
}

main()

File: leetcode_804/solution.go
package main

import (
	"fmt"
)

func uniqueMorseRepresentations(words []string) int {
	codes := []string{
		".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", 
		"-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."}
	mapset := make(map[string]bool)
	for _, word := range words {
		var buf string
		for _, c := range(word) {
			buf += codes[c - 'a']
		}
		mapset[buf] = true
	}
	return len(mapset)
}

func main () {
	words := []string{"gin", "zen", "gig", "msg"}
	ret := uniqueMorseRepresentations(words)
	fmt.Println(ret)
}